File: built-in plugins dsp/mb-pad.cpp
/**
 * MB Synth Pad
 * Category : instrument
 * Type     : pad
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Atmospheric pad synthesizer with rich textures
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PAD_H
#define MB_PAD_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPad : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-pad";
    static constexpr const char* PLUGIN_NAME    = "MB Synth Pad";
    static constexpr const char* PLUGIN_TYPE    = "pad";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float detune = 0.5f;  // range [0, 1]
    float filter = 0.5f;  // range [0, 1]
    float chorus = 0.4f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbPad() = default;
    ~MbPad() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.detune = std::clamp(params.detune, 0f, 1f);
        params.filter = std::clamp(params.filter, 0f, 1f);
        params.chorus = std::clamp(params.chorus, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Synth Pad
        return input;
    }
};

#endif // MB_PAD_H

File: maxbooster7.5/built-in plugins dsp/mb-pad.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbPadPlugin: PluginDefinition = {
  id: "mb-pad",
  slug: "mb-pad",
  name: "MB Synth Pad",
  category: "instrument",
  type: "pad",
  version: "1.0.0",
  description: "Atmospheric pad synthesizer with rich textures",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: -7, gain: 0.25 },
    { type: "sawtooth", detune: 7, gain: 0.25 },
    { type: "sine", detune: 1200, gain: 0.2 },
    { type: "triangle", detune: 0, gain: 0.3 },
  ],
  envelope: { attack: 0.8, decay: 1.0, sustain: 0.9, release: 2.0 },
  parameters: [
    {
      id: "detune",
      name: "Detune",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "filter",
      name: "Filter",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "chorus",
      name: "Chorus",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { detune: 0.5, filter: 0.5, chorus: 0.4, volume: 0.7 },
};

export default MbPadPlugin;
